src/TargetDescription.cc

```diff
@@ -0,0 +1,124 @@
+  const char* arch_prefix{nullptr};
+
+public:
+  std::string result() noexcept { return stream.str(); }
+
+  template <typename Any>
+  friend FeatureStream& operator<<(FeatureStream& stream, Any any) noexcept;
+};
+
+template <typename Any>
+FeatureStream& operator<<(FeatureStream& stream, Any any) noexcept {
+  stream.stream << any;
+  return stream;
+}
+
+template <>
+FeatureStream& operator<<(FeatureStream& stream,
+                          rr::SupportedArch arch) noexcept {
+  stream << "<architecture>";
+  switch (arch) {
+    case rr::x86:
+      stream << "i386";
+      stream.arch_prefix = "32bit-";
+      break;
+    case rr::x86_64:
+      stream << "i386:x86-64";
+      stream.arch_prefix = "64bit-";
+      break;
+    case rr::aarch64:
+      stream << "aarch64";
+      stream.arch_prefix = "aarch64-";
+      break;
+  }
+  stream << "</architecture>\n";
+  return stream;
+}
+
+template <>
+FeatureStream& operator<<(FeatureStream& stream,
+                          TargetFeature feature) noexcept {
+  DEBUG_ASSERT(stream.arch_prefix != nullptr &&
+               "No architecture has been provided to description");
+  stream << R"(  <xi:include href=")" << stream.arch_prefix;
+  switch (feature) {
+    case TargetFeature::Core:
+      stream << "core.xml";
+      break;
+    case TargetFeature::Linux:
+      stream << "linux.xml";
+      break;
+    case TargetFeature::SSE:
+      stream << "sse.xml";
+      break;
+    case TargetFeature::AVX:
+      stream << "avx.xml";
+      break;
+    case TargetFeature::PKeys:
+      stream << "pkeys.xml";
+      break;
+    case TargetFeature::Segment:
+      stream << "seg.xml";
+      break;
+  }
+  stream << R"("/>)" << '\n';
+  return stream;
+}
+
+TargetDescription::TargetDescription(rr::SupportedArch arch,
+                                     u32 cpu_features) noexcept
+    : arch(arch), target_features() {
+
+  // default-assumed registers per arch
+  switch (arch) {
+    case rr::x86:
+      target_features.push_back(TargetFeature::Core);
+      target_features.push_back(TargetFeature::SSE);
+      target_features.push_back(TargetFeature::Linux);
+      break;
+    case rr::x86_64:
+      target_features.push_back(TargetFeature::Core);
+      target_features.push_back(TargetFeature::SSE);
+      target_features.push_back(TargetFeature::Linux);
+      target_features.push_back(TargetFeature::Segment);
+      break;
+    case rr::aarch64:
+      target_features.push_back(TargetFeature::Core);
+      break;
+  }
+
+  if (cpu_features & rr::GdbServerConnection::CPU_AVX) {
+    DEBUG_ASSERT((arch == rr::x86 || arch == rr::x86_64) && "unexpected arch");
+    target_features.push_back(TargetFeature::AVX);
+  }
+
+  if (cpu_features & rr::GdbServerConnection::CPU_PKU) {
+    DEBUG_ASSERT((arch == rr::x86 || arch == rr::x86_64) && "unexpected arch");
+    target_features.push_back(TargetFeature::PKeys);
+  }
+}
+
+static constexpr auto Header = R"(<?xml version="1.0"?>
```

Comment:
Don't need to make this `constexpr` or `auto`. Also, name should be `header` or possibly `HEADER`. E.g. use
`static const char header[] = ...`

Reply:
... or else just write it inline, that's fine

Reply:
Changed it to `header`, but refrained from inlining it directly as raw string literals make code look weird, though that's a personal taste though, and I'll gladly change if it you like.